### PR TITLE
Fix compile with ancient libcurl versions

### PR DESCRIPTION
**Pull request by drons**

### CODE REVIEW

**Comment by rouault** [comment]
Ubuntu 10.04 (so 7 years ago) ship with curl 7.19.7 . Which system do you want to support that would have an older curl library ?

**Comment by drons** [comment]
Unfortunately we still support systems with the ancient version of Red Hat.

**Comment by rouault** [comment]
RHEL 5 ? I thought it isn't supported anymore by RH. Anyway at some point GDAL will require C++11.
I don't see much value for the upstream project to support building on such ancient systems. Of course you're still free to use your patch for your own builds.

**Comment by drons** [comment]
Yes, we have RHEL 5 fork - not a native RHEL.
I am think that my patch does not break anything and does not complicate code maintenance.
Only some clarifications related to the libcurl version.

**Comment by rouault** [comment]
@schwehr Any opinion on this ? GH suggests you as reviewer ;-)

File: gdal/port/cpl_http.cpp [hunk 1]
```diff
@@ -758,27 +758,44 @@ void* CPLHTTPSetOptions(void *pcurl, const char * const* papszOptions)
     }
 #endif
 
-    // Enable following redirections.  Requires libcurl 7.10.1 at least.
+    // Enable following redirections.  Requires libcurl 7.19.1 at least.
+#if LIBCURL_VERSION_NUM >= 0x071301
```
Comment (schwehr) [inline gdal/port/cpl_http.cpp]: This version doesn't match the comment.

https://github.com/curl/curl/blob/master/include/curl/curlver.h#L37

Says that it's comparing against 7.13.1.

File: gdal/port/cpl_http.cpp [hunk 2]
```diff
@@ -758,27 +758,44 @@ void* CPLHTTPSetOptions(void *pcurl, const char * const* papszOptions)
...
     curl_easy_setopt(http_handle, CURLOPT_FOLLOWLOCATION, 1 );
     curl_easy_setopt(http_handle, CURLOPT_MAXREDIRS, 10 );
     curl_easy_setopt(http_handle, CURLOPT_POSTREDIR, CURL_REDIR_POST_ALL );
+#endif//LIBCURL_VERSION_NUM >= 0x071301
```
Comment (schwehr) [inline gdal/port/cpl_http.cpp]: Spaces please.  2 before the // and one after.

`#endif  // LIBCURL_VERSION_NUM >= 0x071301`

File: gdal/port/cpl_http.cpp [hunk 3]
```diff
@@ -758,27 +758,44 @@ void* CPLHTTPSetOptions(void *pcurl, const char * const* papszOptions)
...
 
+    // CURLOPT_CONNECTTIMEOUT_MS is defined in curl 7.16.2 or newer.
```
Comment (schwehr) [inline gdal/port/cpl_http.cpp]: Nowhere do I see something that ensures that the version will be >= `0x071602`

File: gdal/port/cpl_http.cpp [hunk 4]
```diff
@@ -758,27 +758,44 @@ void* CPLHTTPSetOptions(void *pcurl, const char * const* papszOptions)
...
         CSLFetchNameValue( papszOptions, "CONNECTTIMEOUT" );
     if( pszConnectTimeout == NULL )
         pszConnectTimeout = CPLGetConfigOption("GDAL_HTTP_CONNECTTIMEOUT", NULL);
     if( pszConnectTimeout != NULL )
+    {
+#if LIBCURL_VERSION_NUM >= 0x071002
         curl_easy_setopt(http_handle, CURLOPT_CONNECTTIMEOUT_MS,
                          static_cast<int>(1000 * CPLAtof(pszConnectTimeout)) );
+#else
+        curl_easy_setopt(http_handle, CURLOPT_CONNECTTIMEOUT,
```
Comment (schwehr) [inline gdal/port/cpl_http.cpp]: What are the units of `CURLOPT_CONNECTTIMEOUT`?

Add a comment.

File: gdal/port/cpl_http.cpp [hunk 5]
```diff
@@ -758,27 +758,44 @@ void* CPLHTTPSetOptions(void *pcurl, const char * const* papszOptions)
...
     const char *pszTimeout = CSLFetchNameValue( papszOptions, "TIMEOUT" );
     if( pszTimeout == NULL )
         pszTimeout = CPLGetConfigOption("GDAL_HTTP_TIMEOUT", NULL);
     if( pszTimeout != NULL )
+    {
+#if LIBCURL_VERSION_NUM >= 0x071002
         curl_easy_setopt(http_handle, CURLOPT_TIMEOUT_MS,
                          static_cast<int>(1000 * CPLAtof(pszTimeout)) );
+#else
+        curl_easy_setopt(http_handle, CURLOPT_TIMEOUT,
```
Comment (schwehr) [inline gdal/port/cpl_http.cpp]: Units of `CURLOPT_TIMEOUT`?  Add a comment.

**Comment by rouault** [comment]
@schwehr My question was more if it made sense to support such ancient versions of libcurl

File: gdal/port/cpl_http.cpp (hunk 1 above)
Comment (rouault) [inline gdal/port/cpl_http.cpp]: 0x13 = 19

**Comment by schwehr** [comment]
I think this patch is fine for any of the older branches of GDAL.

However...

For trunk (and maybe 2.2), we should try to avoid patches like this.  Each time we add something like this, it seems like no big deal, but it really adds up across all of GDAL especially w.r.t. testing and any need for future refactoring to add capabilities.

```
find . -name \*.cpp -o -name \*.h | xargs egrep '#if' | egrep -v 'notdef|if 0|DEBUG|VERBOSE' | wc -l
7942
```

Heh... If I'd read one line more upstream, I'd have seen that the numbers are indeed actually hex.

**Comment by drons** [comment]
It's amazing, but the numbers are really [HEX](https://github.com/curl/curl/blob/master/include/curl/curlver.h#L60)
Units are seconds: [TIMEOUT](https://github.com/curl/curl/blob/master/include/curl/curl.h#L926), [CONNECTTIMEOUT](https://github.com/curl/curl/blob/master/include/curl/curl.h#L1131)
It is not very good to use `#ifdef`, but this is the fee for a cross-platform

**Comment by schwehr** [comment]
Agreed that we have use `#ifdef`, it's just that we also want to keep the complexity from getting out of control.  It's a tough balancing act of two competing worthy goals.

**Comment by rouault** [comment]
My main point is that supporting libcurl versions that are more than 10 years old seems overkill. I suppose even the RHEL 5 clones aren't longer security maintained given that RH doesn't do it anymore for the base RHEL 5 (https://en.wikipedia.org/wiki/CentOS confirms it, https://www.scientificlinux.org/downloads/sl-versions/sl5/ too ). You can also build a newer libcurl if needed, or disable curl support when building GDAL

**Comment by drons** [comment]
I also do not want to support the ancient distribution versions, but we have a very conservative customer.
I think that the new libcurl == it's new dependencies, and this is impossible
Without libcurl, the meaning of our application is lost.

I'm think we need to create GDAL dependencies white list with versions for next releases.

**Comment by schwehr** [comment]
Those of us who have to support very unusual environments sometimes have to maintain some patches.  I have a bunch of patches that I need that don't make sense in the tree.  It's definitely more work.

**Comment by rouault** [comment]
Closing as wontfix

**Pull request closed without merging**